# Time dilation and length contraction

Many amazing aspects of special relativity, including *time dilation* and *length contraction*, can be understood by using the *Lorentz factor*. It expresses how much the measurements of time, length, and other physical properties change for an object while it moves. The Lorentz factor is given by:

$$\gamma = \frac{1}{\sqrt{1 - v^2/c^2}}$$

where $v$ is the relative velocity between the two bodies and $c$ is the speed of light. The Lorentz factor is always greater than or equal to 1, and it approaches infinity as the relative velocity approaches the speed of light.

### Time dilation

Time dilation is when time slows down for an object in motion relative to an observer. The slowed duration is given by **multiplying the normal duration by the Lorentz factor**. Or in more mathematical terms:

$$\Delta t' = \gamma \Delta t$$

So given a Lorentz factor of 10 (equivalent to a relative speed of about 298,300 km every second), 1 second becomes 10 seconds for an observer.

### Length contraction

Length contraction is when an object in motion shortens relative to an observer. The contracted length is given by **dividing the normal length by the Lorentz factor**. Or in more mathematical terms:

$$L' = L/\gamma$$

So given a Lorentz factor of 10, a 1 metre rod becomes 0.1 metres (10 cm) for an observer.

## Putting it into practice

Let's work it out the Lorentz factor using Python:

In [1]:
import math

c = 299_792_458.0  # speed of light, metres per second
c_squared = c**2  # c squared


def lorentz_factor(v: float) -> float:
    if v >= c:
        return math.inf
    return 1.0 / math.sqrt(1.0 - (v**2 / c_squared))


lorentz = lorentz_factor(0.9 * c)  # 90% of the speed of light
print(f"Lorentz factor for 0.9c = {lorentz:.2f}")
print(f"1 second becomes {1.0 * lorentz:.2f} seconds")
print(f"1 metre becomes {1.0 / lorentz:.2f} metres")

Lorentz factor for 0.9c = 2.29
1 second becomes 2.29 seconds
1 metre becomes 0.44 metres


What does this mean in practice? Let's imagine a ship is travelling at 0.9c relative to the Earth. If we watch it using a telescope:

- Time on board slows down, so 1 second on ship takes 2.29 seconds on Earth. To us, the ship appears to be moving in slow motion.
- In the direction of travel, lengths get contracted so a 1 metre ruler on ship appears to be only 44cm long from the Earth. To us, the ship appears to be squashed.

## Symmetry

Interestingly this effect is completely symmetrical. If someone on the ship uses a telescope to look back at Earth, they will see the same slowing and length contraction! But of course their own length and time flow appear perfectly normal to them.

Since the ship seems slowed from Earth, and the Earth seems slowed from the ship, they will disagree about the time of any events they both witness. It's even possible for them to disagree on the order of events, depending on the exact circumstances.

## Length contraction for interstellar travel

Length contraction is also symmetric, so a moving body sees the rest of the universe shrink in its direction of travel. Let’s assume you are on a ship travelling at 0.9c between Earth and Proxima 4.2 light years away. We know the Lorentz factor is 2.29, so that means the distance to Proxima becomes 4.2/2.29 = **1.83 light years**. It actually gets closer!

Since you are travelling at 0.9c it will take 1.834 / 0.9 = **2.04 years** to make the journey. You have travelled over 4 light years in only 2 years, although you’ve never actually exceeded the speed of light.

Using this effect, you can cross the entire universe in a human lifetime, see my [tour of the universe notebook](https://github.com/lookbusy1344/Relativity/blob/main/Python/Universe.ipynb).

Now let’s look at some larger Lorentz factors.

## A table of Lorentz factors

In [2]:
from prettytable import PrettyTable, TableStyle
from IPython.display import Markdown

table = PrettyTable()
table.set_style(TableStyle.MARKDOWN)
table.align = "r"
table.field_names = [
    "Velocity",
    "Lorentz factor",
    "Time dilation",
    "Length contraction",
]


def process(fraction: float) -> None:
    lorentz = lorentz_factor(fraction * c)
    table.add_row(
        [
            f"{fraction}c",
            f"{lorentz:.3f}",
            f"{1.0 * lorentz:.2f} sec",
            f"{1.0 / lorentz * 100.0:.2f} cm",
        ]
    )


# Edit these values to see the effect of different velocities
process(0.1)
process(0.2)
process(0.4)
process(0.6)
process(0.7)
process(0.8)
process(0.9)
process(0.99)
process(0.999)
process(0.9999)
process(0.99999)
process(0.9999999)
# process(1.0)

print("Time dilation is the amount 1 second is slowed to")
print("Length contraction is the amount 1 metre is contracted to")
markdown_table = table.get_string()
display(Markdown(markdown_table))

Time dilation is the amount 1 second is slowed to
Length contraction is the amount 1 metre is contracted to


|   Velocity | Lorentz factor | Time dilation | Length contraction |
|----------: |--------------: |-------------: |------------------: |
|       0.1c |          1.005 |      1.01 sec |           99.50 cm |
|       0.2c |          1.021 |      1.02 sec |           97.98 cm |
|       0.4c |          1.091 |      1.09 sec |           91.65 cm |
|       0.6c |          1.250 |      1.25 sec |           80.00 cm |
|       0.7c |          1.400 |      1.40 sec |           71.41 cm |
|       0.8c |          1.667 |      1.67 sec |           60.00 cm |
|       0.9c |          2.294 |      2.29 sec |           43.59 cm |
|      0.99c |          7.089 |      7.09 sec |           14.11 cm |
|     0.999c |         22.366 |     22.37 sec |            4.47 cm |
|    0.9999c |         70.712 |     70.71 sec |            1.41 cm |
|   0.99999c |        223.607 |    223.61 sec |            0.45 cm |
| 0.9999999c |       2236.068 |   2236.07 sec |            0.04 cm |

Note that objects with mass cannot travel at the speed of light, and the Lorentz factor is infinite. At $c$ the equation breaks down, with time stopping, lengths contracting to zero, and mass/energy becoming infinite.

## It's all relative

These effects are all **relative**, not absolute. There is no universal background to measure your speed against, and an object can appear to have different speeds from different perspectives (frames of reference).

Let's recalculate using an object travelling at 0.999c relative to the Earth, but observed from 6 different frames of reference, each traveling at a different speed:

In [3]:
table = PrettyTable()
table.set_style(TableStyle.MARKDOWN)
table.align = "r"
table.field_names = [
    "Velocity of observer",
    "Velocity of object",
    "Relative velocity",
    "Lorentz factor",
    "Time dilation",
    "Length contraction",
]


def processr(v1: float, v2: float) -> None:
    f = math.fabs(v1 - v2)
    lorentz = lorentz_factor(f * c)
    table.add_row(
        [
            f"{v1}c",
            f"{v2}c",
            f"{f:.3f}c",
            f"{lorentz:.6f}",
            f"{1.0 * lorentz:.5f} sec",
            f"{1.0 / lorentz * 100.0:.5f} cm",
        ]
    )


# Edit these values to see the effect of different velocities
processr(0.0, 0.999)
processr(0.1, 0.999)
processr(0.5, 0.999)
processr(0.9, 0.999)
processr(0.99, 0.999)
processr(0.999, 0.999)

print("Velocity of observer - velocity of observer relative to Earth")
print("Velocity of object - velocity of object relative to Earth")
print(
    "Relative velocity - the difference between the two speeds, which defines the lorentz factor"
)
markdown_table = table.get_string()
display(Markdown(markdown_table))

Velocity of observer - velocity of observer relative to Earth
Velocity of object - velocity of object relative to Earth
Relative velocity - the difference between the two speeds, which defines the lorentz factor


| Velocity of observer | Velocity of object | Relative velocity | Lorentz factor | Time dilation | Length contraction |
|--------------------: |------------------: |-----------------: |--------------: |-------------: |------------------: |
|                 0.0c |             0.999c |            0.999c |      22.366272 |  22.36627 sec |         4.47102 cm |
|                 0.1c |             0.999c |            0.899c |       2.283373 |   2.28337 sec |        43.79486 cm |
|                 0.5c |             0.999c |            0.499c |       1.153932 |   1.15393 sec |        86.66020 cm |
|                 0.9c |             0.999c |            0.099c |       1.004937 |   1.00494 sec |        99.50874 cm |
|                0.99c |             0.999c |            0.009c |       1.000041 |   1.00004 sec |        99.99595 cm |
|               0.999c |             0.999c |            0.000c |       1.000000 |   1.00000 sec |       100.00000 cm |

## One object in 6 frames of reference

Notice that all the rows describe the same object simultaneously from 6 different frames of reference. This is important! An object travelling at 0.999c relative to the Earth has 6 different Lorentz factors from 6 different frames of reference, simultaneously.

Any consequences of the lorentz factor (including time dilation, length contraction, and relativistic mass/energy) are strictly frame dependent. They are not absolute changes that apply from all frames of reference.

We haven't talked much about relativistic mass and energy, but [it's discussed in detail here](https://github.com/lookbusy1344/Relativity/blob/main/Python/Relativistic%20mass.ipynb)

More about the Lorentz factor here:

https://en.wikipedia.org/wiki/Lorentz_factor